sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/BeamFnStateGrpcClientCache.java

```diff
@@ -0,0 +1,173 @@
+
+  private final ConcurrentMap<ApiServiceDescriptor, BeamFnStateClient> cache;
+  private final Function<ApiServiceDescriptor, ManagedChannel> channelFactory;
+  private final BiFunction<Function<StreamObserver<StateResponse>,
+      StreamObserver<StateRequest>>,
+      StreamObserver<StateResponse>,
+      StreamObserver<StateRequest>> streamObserverFactory;
+  private final PipelineOptions options;
+  private final Supplier<String> idGenerator;
+
+  public BeamFnStateGrpcClientCache(
+      PipelineOptions options,
+      Supplier<String> idGenerator,
+      Function<BeamFnApi.ApiServiceDescriptor, ManagedChannel> channelFactory,
+      BiFunction<Function<StreamObserver<StateResponse>, StreamObserver<StateRequest>>,
+          StreamObserver<StateResponse>,
+          StreamObserver<StateRequest>> streamObserverFactory) {
+    this.options = options;
+    this.idGenerator = idGenerator;
+    this.channelFactory = channelFactory;
+    this.streamObserverFactory = streamObserverFactory;
+    this.cache = new ConcurrentHashMap<>();
+  }
+
+  /**(
+   * Creates or returns an existing {@link BeamFnStateClient} depending on whether the passed in
+   * {@link ApiServiceDescriptor} currently has a {@link BeamFnStateClient} bound to the same
+   * channel.
+   */
+  public BeamFnStateClient forApiServiceDescriptor(ApiServiceDescriptor apiServiceDescriptor)
+      throws IOException {
+    return cache.computeIfAbsent(apiServiceDescriptor, this::createBeamFnStateClient);
+  }
+
+  private BeamFnStateClient createBeamFnStateClient(ApiServiceDescriptor apiServiceDescriptor) {
+    return new GrpcStateClient(apiServiceDescriptor);
+  }
+
+  /**
+   * A {@link BeamFnStateClient} for a given {@link ApiServiceDescriptor}.
+   */
+  private class GrpcStateClient implements BeamFnStateClient {
+    private final ApiServiceDescriptor apiServiceDescriptor;
+    private final ConcurrentMap<String, CompletableFuture<StateResponse>> outstandingRequests;
+    private final StreamObserver<StateRequest> outboundObserver;
+    private final ManagedChannel channel;
+    private volatile RuntimeException closed;
+
+    private GrpcStateClient(ApiServiceDescriptor apiServiceDescriptor) {
+      this.apiServiceDescriptor = apiServiceDescriptor;
+      this.outstandingRequests = new ConcurrentHashMap<>();
+      this.channel = channelFactory.apply(apiServiceDescriptor);
+      this.outboundObserver = streamObserverFactory.apply(
+          BeamFnStateGrpc.newStub(channel)::state, new InboundObserver());
+    }
+
+    @Override
+    public void handle(
+        StateRequest.Builder requestBuilder, CompletableFuture<StateResponse> response) {
+      requestBuilder.setId(idGenerator.get());
+      StateRequest request = requestBuilder.build();
+      outstandingRequests.put(request.getId(), response);
+
+      // If the server closes, gRPC will throw an error if onNext is called.
+      LOG.debug("Sending StateRequest {}", request);
+      outboundObserver.onNext(request);
```

Comment:
Wrap this and throw a more meaningful error?

Reply:
Also why not check for `closed`?

Reply:
That way I think you could also avoid copying the map.

Reply:
There is a race condition of checking `closed` unless I synchronize on updating the map which I don't want to have to do.

Lets say the code was like this
```
if (closed) {
  throw Exception;
}
updateMap
sendRequest
```

Lets say thread T1 was trying to send a new request while thread T2 was trying to close.
Even if close gets set, T1 may have already passed the point where it checked if closed was set and it may still insert something into the requests map but will fail when it invokes onNext.

T2 needs to make a copy since it needs to iterate over a map which may still be inserted into or iteration risks throwing a ConcurrentModificationException.

Finally, I believe that rewrapping the exception should really happen much higher closer to the users code saying something like "Failed to append/clear value state because blah failed" without adding to a potentially long chain of causes. This BeamFnStateClient doesn't have the high level context to give a great error message.

Reply:
You are right about the race condition. But since we are already hitting the memory synchronization barrier by using the `ConcurrentHashMap`, I am wondering if `synchronized` may just be fine here, to make it simpler.

Reply:
I mean the entire `handle` method in `synchronized`, and using a regular 'HashMap` instead.

Reply:
Memory synchronization and using synchronized isn't the same thing since one is about when a value becomes visible to another CPU and the synchronized is about using a lock which would serialize all requests / responses coming in/out with modifications to the HashMap.

ConcurrentHashMap allows for parallel modifications because it uses a lock per "sub" map that is partitioned based upon hash code.

Reply:
I know. The memory synchronization comment was if you were worried about costs of it. What I was getting at is that it is all asynchronous operations in the critical path (is it not the case?), so maybe serialization of requests / responses would not be that bad. In my experience, optimizing multi-threaded code for finer grained blocks has lead to bugs that are hard to debug. So I usually prefer coarser blocks unless there is a strong reason not to do so.

Reply:
That is a valid point but for now will leave as is.